src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp

```diff
@@ -104,321 +150,710 @@ enum Spells
 
-        InstanceScript* instance;
-        ObjectGuid MemoryGUID;
-
-        bool bHealth;
-        bool bDone;
-
-        uint32 uiHolyFireTimer;
-        uint32 uiHolySmiteTimer;
-        uint32 uiRenewTimer;
-        uint32 uiResetTimer;
-
-        void Reset() override
+        void OnReset() override
         {
-            me->RemoveAllAuras();
-
+            me->RemoveAura(SPELL_SHIELD);
             Initialize();
+            argent_challenge_baseAI::OnReset();
+        }
 
-            if (Creature* pMemory = ObjectAccessor::GetCreature(*me, MemoryGUID))
-                if (pMemory->IsAlive())
-                    pMemory->RemoveFromWorld();
+        void SummonedCreatureDies(Creature* /*summon*/, Unit* /*killer*/) override {
+            me->InterruptNonMeleeSpells(true);
+            Talk(SAY_MEMORY_DEATH);
+            me->GetMotionMaster()->Clear();
+            me->RemoveAura(SPELL_SHIELD);
         }
 
-        void SetData(uint32 uiId, uint32 /*uiValue*/) override
+        uint32 GetData(uint32 type) const override
         {
-            if (uiId == 1)
-                me->RemoveAura(SPELL_SHIELD);
+            if (type != DATA_MEMORY_ENTRY || !_memory)
+                return 0;
+
+            return _memory->GetEntry();
         }
 
         void DamageTaken(Unit* /*done_by*/, uint32 &damage) override
         {
+            if (instance->GetBossState(DATA_ARGENT_CHALLENGE) == DONE)
+            {
+                damage = 0;
+                return;
+            }
+
+            if (!_memorySummoned && me->HealthBelowPctDamaged(25, damage))
+            {
+                damage = me->GetHealth() - me->CountPctFromMaxHealth(25);
+                _memorySummoned = true;
+                me->InterruptNonMeleeSpells(true);
+                DoCastAOE(SPELL_HOLY_NOVA);
+                Talk(SAY_MEMORY_SUMMON);
+                DoCastAOE(SPELL_SHIELD);
+                DoCastAOE(SPELL_CONFESS);
+                me->AttackStop();
+                events.ScheduleEvent(EVENT_SUMMON_MEMORY, 2 * IN_MILLISECONDS);
+            }
+
             if (damage >= me->GetHealth())
             {
                 damage = 0;
-                EnterEvadeMode();
-                me->SetFaction(FACTION_FRIENDLY);
-                bDone = true;
+                if (instance->GetBossState(DATA_ARGENT_CHALLENGE) != SPECIAL)
+                {
+                    instance->SetBossState(DATA_ARGENT_CHALLENGE, SPECIAL);
+                    events.Reset();
+                    Talk(SAY_DEFEATED_P);
+                }
             }
         }
 
-        void MovementInform(uint32 MovementType, uint32 Point) override
+        void JustSummoned(Creature* summon) override
         {
-            if (MovementType != POINT_MOTION_TYPE || Point != 0)
+            argent_challenge_baseAI::JustSummoned(summon);
+            
+            ObjectGuid _memoryGuid = summon->GetGUID();
+            _memory = ObjectAccessor::GetCreature(*me, _memoryGuid);
```

Comment:
NO - don't do this.

You are not allowed to store any entity pointer between server ticks - units can despawn, or be unloaded for some other reason - then you're dereferencing an invalid pointer, and get some _lovely_ bugs.

Store the GUID and access it using `ObjectAccessor` when you need it - and `nullptr` check the `ObjectAccessor` return!